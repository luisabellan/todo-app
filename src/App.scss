/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/

html,
body,
div,
span,
applet,
object,
iframe,
h1,
h2,
h3,
h4,
h5,
h6,
p,
blockquote,
pre,
a,
abbr,
acronym,
address,
big,
cite,
code,
del,
dfn,
em,
img,
ins,
kbd,
q,
s,
samp,
small,
strike,
strong,
sub,
sup,
tt,
var,
b,
u,
i,
center,
dl,
dt,
dd,
ol,
ul,
li,
fieldset,
form,
label,
legend,
table,
caption,
tbody,
tfoot,
thead,
tr,
th,
td,
article,
aside,
canvas,
details,
embed,
figure,
figcaption,
footer,
header,
hgroup,
menu,
nav,
output,
ruby,
section,
summary,
time,
mark,
audio,
video {
  margin: 0;
  padding: 0;
  border: 0;
  font-size: 100%;
  font: inherit;
  vertical-align: baseline;
}
/* HTML5 display-role reset for older browsers */
article,
aside,
details,
figcaption,
figure,
footer,
header,
hgroup,
menu,
nav,
section {
  display: block;
}
body {
  line-height: 1;
}
ol,
ul {
  list-style: none;
}
blockquote,
q {
  quotes: none;
}
blockquote:before,
blockquote:after,
q:before,
q:after {
  content: '';
  content: none;
}
table {
  border-collapse: collapse;
  border-spacing: 0;
}

html,
body {
  background-color: #bce6ab;
  margin: 0;
  .app-wrapper {
    background-color: #bce6ab;
    display: flex;
    width: 100%;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;

    h2 {
      color: rgb(29, 138, 93);
      font-size: 2rem;
      margin: 3rem auto;
      text-align: center;
    }

    form {
      display: flex;
      flex-direction: row;
      justify-content: flex-start;
      align-content: space-around;
      width: 100%;
      input {
        width: 20rem;
        text-align: start;
        margin-left: 2rem;
        font-size: 1.5rem;
        background-color: #b5f5a5;
        color: rgb(9, 19, 14);
        margin-top: 2rem;
        height: 2rem;
        border-radius: 1rem;
        outline: none;
      }

      #buttons {
        display: flex;
        flex-direction: row;
        justify-content: center;

        #button-left {
          display: flex;

          button.add-to-do-btn {
            display: flex;
            flex-direction: row;
            align-items: center;
            margin-top: 2rem;
            height: 2.3rem;
            width: 5rem;
            border-radius: 0.2rem;

            span {
              margin: 0 auto;
            }
          }
        }
        #button-right {
          button.clear-btn {
            display: flex;
            flex-direction: row;
            align-items: center;
            align-self: center;
            margin-top: 2rem;
            height: 2.3rem;
            width: 5rem;
            border-radius: 0.2rem;
            span {
              margin: 0 auto;
            }
          }
        }
      }
    }

    .todolist-wrapper {
      background-color: rgb(29, 138, 93);
      display: flex;
      flex-direction: column;
      justify-content: flex-start;
      align-self: center;
      width: 25rem;
      height: 20rem;
      padding-top: 2rem;
      margin: 0 auto;
      margin-top: 4rem;
      overflow-y: auto;
      scrollbar-color: dark;
      scroll-behavior: smooth;
      border-radius: 0.2rem;

      input {
        margin-left: 3rem;
      }

      span {
        font-size: 1.3rem;
        color: #f8faf8;
        font-weight: 600;
        height: 5vh;
        margin: 0 auto;
        text-align: center;
        margin: 0 auto;
        margin-top: 2rem;
        margin-left: 1.2rem;
        text-align: center;
        padding-top: 2rem;
      }
    }
    .item {
      cursor: pointer;
    }

    .item.completed {
      text-decoration: solid line-through red;
    }
    footer {
      position: static;
      padding: 2rem;
      bottom: 1rem;
      margin-top: 5rem;

      .copyright {
        color: rgba(6, 92, 72, 0.863);
      }
    }
  }
}
