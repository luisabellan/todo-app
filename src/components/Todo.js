import React from "react";

export default function Todo(props) {
  return (
    <div
      className={`item ${props.item.completed ? " completed" : ""}`}
      onClick={() => props.toggleItem(props.item.id)}
    >
      <p style={{ fontFamily: "Indie Flower", fontSize: "2rem" }}>
        {props.item.name}
      </p>
    </div>
  );
}
