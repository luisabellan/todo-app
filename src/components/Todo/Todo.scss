.task {
  display: flex;
  font-family: "Indie Flower";
  input {
    margin-right: 1rem;
  }
  span {
    font-size: "3rem";
  }
}
