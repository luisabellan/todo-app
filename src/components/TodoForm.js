import React, { useState } from "react";

export default function TodoForm(props) {
  const initialState = {
    todoItem1: "",
  };
  const [state, setState] = useState(initialState);

  const handleChanges = (e) => {
    // update state with each keystroke
    // console.log(e)
    setState({ [e.target.name]: e.target.value });
    // console.log(state);
  };

  // class property to submit form
  const submitItem = (e) => {
    e.preventDefault();
    //console.log(props);
    props.addItem(e, state.todoItem1);
    setState({ todoItem1: "" });
  };

  // console.log('rendering form', state.todoItem);

  return (
    <>
      <form onSubmit={submitItem}>
        <h2>New Task</h2>
        <input
          type="text"
          data-testid="todo-input"
          value={state.todoItem1}
          name="todoItem1"
          onChange={handleChanges}
          placeholder="New task"
        />
        /
        <button
          className="add-to-do-btn"
          style={{
            fontFamily: "Jomhuria,sans-serif",
            outline: "none",
            fontSize: "3.2rem",
          }}
        >
          <span>Add</span>
        </button>
        <button
          className="clear-btn"
          onClick={props.clearCompleted}
          style={{
            fontFamily: "Jomhuria,sans-serif",
            outline: "none",
            fontSize: "3.2rem",
          }}
        >
          <span>Clean</span>
        </button>
      </form>
    </>
  );
}
