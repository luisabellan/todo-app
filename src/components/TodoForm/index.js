import React, { useState } from 'react';
import PropTypes from 'prop-types';

TodoForm.propTypes = {
  addItem: PropTypes.func,
  clearCompleted: PropTypes.func
};

export default function TodoForm(props) {
  const initialState = {
    todoItem1: ''
  };
  const [state, setState] = useState(initialState);

  const handleChanges = (e) => {
    // update state with each keystroke
    // console.log(e)
    setState({ [e.target.name]: e.target.value });
    // console.log(state);
  };

  // class property to submit form
  const submitItem = (e) => {
    e.preventDefault();
    //console.log(props);
    props.addItem(e, state.todoItem1);
    setState({ todoItem1: '' });
  };

  // console.log('rendering form', state.todoItem);

  return (
    <>
      <form onSubmit={submitItem}>
        <input
          type="text"
          data-testid="todo-input"
          value={state.todoItem1}
          placeholder="New Task"
          name="todoItem1"
          onChange={handleChanges}
        ></input>

        <button className="add-to-do-btn">
          <span>Add</span>
        </button>
        <button className="clear-btn" onClick={props.clearCompleted}>
          <span>Clean</span>
        </button>
      </form>
    </>
  );
}
