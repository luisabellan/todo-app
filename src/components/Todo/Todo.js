/* eslint-disable no-labels */
/* eslint-disable no-unused-expressions */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './Todo.scss';

Todo.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string.isRequired,
    completed: PropTypes.bool,
  }),

  toggleItem: PropTypes.func,
  handleChanges: PropTypes.func
};

export default function Todo(props) {

  const todoInitialState = {
    id: "",
    name: "",
    completed: false
  }

  const [todo, setTodo] = useState(todoInitialState)


  return (
    <div className="checkbox" data-testid="todo">
      <input
        type="checkbox"
        data-testid={`checkbox ${props.todo.name}`}
        name={`item ${props.todo.name}`}
        value={props.todo.name}
        onClick={() => props.toggleItem(props.todo.id)}
        onChange={props.handleChanges}
        defaultChecked={props.todo.completed}
      ></input>
      <span data-testid="todo-output">{props.todo.name}</span>
    </div >
  );
}
