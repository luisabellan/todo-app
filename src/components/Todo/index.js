/* eslint-disable no-labels */
/* eslint-disable no-unused-expressions */
import React from "react";
import PropTypes from "prop-types";
import "./Todo.scss";

/* item: PropTypes.exact({
  id: PropTypes.number.isRequired,
  name: PropTypes.string,
  completed: PropTypes.bool,
}); */

Todo.propTypes = {
  item: PropTypes.object,
  toggleItem: PropTypes.func,
  handleChanges: PropTypes.func,
};

export default function Todo(props) {
  return (
    <div
      /* className={`item ${props.item.completed ? " completed" : ""}`}  */
      className="task"
      /*onClick={() => props.toggleItem(props.item.id)}*/
    >
      <input
        type="checkbox"
        name={`item ${props.item.name}`}
        value={props.item.name}
        onClick={() => props.toggleItem(props.item.id)}
        onChange={props.handleChanges}
        defaultChecked={props.item.completed}
      ></input>
      <span>{props.item.name}</span>
    </div>
  );
}
