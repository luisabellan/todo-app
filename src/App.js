import React, { useState, useEffect } from 'react';
import TodoList from './components/TodoList/TodoList';
import TodoForm from './components/TodoForm/TodoForm';
import Header from './components/Header/Header';
import Footer from './components/Footer/Footer';

import './App.scss';

function App() {
  let todoItems = [];

  const [todos, setTodos] = useState(todoItems);

  useEffect(() => {
    if (localStorage.getItem('todos')) {
      setTodos(JSON.parse(localStorage.getItem('todos')));
    }
  }, []);



  // saves todo on local storage
  const saveData = (newTodos) => {

    localStorage.setItem('todos', JSON.stringify(newTodos));
  }


  // Class methods to update state
  const addItem = (e, todo) => {
    e.preventDefault();

    const newItem = {
      //todo item ${item.note}`,
      id: Date.now().toString(),
      note: todo,
      completed: false
    };
    let todoItems = [...todos, newItem];

    /* 
      /\S+/.test(item) 
      if 0 or more whitespace characters only item will not be added       
    */

    if (/\S+/.test(todo)) {
      setTodos(todoItems);
      saveData(todoItems);

    }
  };



  const toggleItem = (todoId) => {
    // console.log(itemId);

    setTodos(
      todos.map((todo) => {
        // console.log(item);
        if (todoId === todo.id) {
          return {
            ...todo,
            completed: !todo.completed
          };
        }

        return todo;
      })
    );
  };
  const clearCompleted = (e) => {
    e.preventDefault();
    let todoItems = todos.filter((item) => item.completed === false);
    setTodos(todoItems);
    saveData(todoItems);
  };
  const clearAll = (e) => {
    e.preventDefault();
    let todoItems = []
    setTodos(todoItems);
    saveData(todoItems);
  };



  return (
    <div className="app-wrapper" data-testid="app">

      <Header className="header" />
      <div className="main">
        <TodoForm
          className="todoform"
          addItem={addItem}
          clearCompleted={clearCompleted}
          clearAll={clearAll}
        />
        {/* {console.log(todos)} */}
        <TodoList style={{ fontFamily: 'Indie Flower, monospaced' }} data-testid="todolist" todos={todos} className="todolist" toggleItem={toggleItem} />
      </div >
      <Footer />
    </div >
  )
}

export default App;
