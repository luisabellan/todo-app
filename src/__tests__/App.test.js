import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from '../App';
import TodoList from '../components/TodoList/TodoList';
import Todo from '../components/Todo/Todo';
import TodoForm from '../components/TodoForm/TodoForm';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';

describe('App', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  /*   beforeEach(() => {
    let todos = [
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy strawberries', completed: false }
    ];
    localStorage.setItem('todos', JSON.stringify(todos));
    render(
      <App>
        <TodoList todos={todos}>
          <Todo />
        </TodoList>
      </App>
    );
  });
 */
  it('renders:title, Text input field, Add button, Delete button and footer', () => {
    let todos = [
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy strawberries', completed: false }
    ];
    localStorage.setItem('todos', JSON.stringify(todos));
    render(
      <App>
        <TodoList todos={todos} />
      </App>
    );
    const title = screen.getByText(/Todo App/);
    const placeholder = screen.getByPlaceholderText(/New Task/);
    const addTodoButton = screen.getByText(/Add/);
    const cleanTodoButton = screen.getByText(/Done/);
    const clearTodoButton = screen.getByText(/Delete/);
    const footer = screen.getByText(/Made with ðŸ’“ and â˜• by/);
    const footerLink = screen.getByText(/Luis Abellan/);

    expect(title).toBeInTheDocument();
    expect(placeholder).toBeInTheDocument();
    expect(addTodoButton).toBeInTheDocument();
    expect(cleanTodoButton).toBeInTheDocument();
    expect(clearTodoButton).toBeInTheDocument();
    expect(footer).toBeInTheDocument();
    expect(footerLink).toBeInTheDocument();
  });

  // Clear item
  it('clear completed', () => {
    let todos = [
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy potatoes', completed: false }
    ];
    /* let todo = { id: 'a', note: 'buy bananas', completed: false }; */

    localStorage.setItem('todos', JSON.stringify(todos[0]));
    render(
      <I18nProvider i18n={i18n}>
        <App>
          <TodoForm />
          <TodoList todos={todos}>
            <Todo todo={todos[0]}></Todo>
          </TodoList>
        </App>
      </I18nProvider>
    );
    /* render(<Todo todo={todos[0]} />);
    render(<TodoForm />); */
    let checkbox = screen.getByTestId(`checkbox ${todos[0].note}`);

    expect(JSON.parse(localStorage.getItem('todos'))).toStrictEqual(
      {
        id: 'a',
        note: 'buy bananas',
        completed: false
      },
      { id: 'b', note: 'buy potatoes', completed: false }
    );

    userEvent.click(checkbox);

    const cleanButton = screen.getByText('Done');
    userEvent.click(cleanButton);

    expect(JSON.parse(localStorage.getItem('todos'))).toStrictEqual({});
    userEvent.click(checkbox);
    expect(checkbox).toBeChecked();

    localStorage.clear();
  });

  // Clear all
  test('clear all', () => {
    let todos = [
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy strawberries', completed: false }
    ];
    localStorage.setItem('todos', JSON.stringify(todos));
    render(
      <I18nProvider i18n={i18n}>
        <App>
          <TodoList todos={todos} />
        </App>
      </I18nProvider>
    );
    localStorage.setItem('todos', JSON.stringify(todos));
    const deleteButton = screen.getByText('Delete');
    expect(deleteButton).toBeInTheDocument();

    expect(JSON.parse(localStorage.getItem('todos'))).toStrictEqual([
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy strawberries', completed: false }
    ]);

    userEvent.click(deleteButton);
    expect(JSON.parse(localStorage.getItem('todos'))).toStrictEqual([]);
    localStorage.clear();
  });

  // Add item
  it('adds item', () => {
    let todos = [
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy strawberries', completed: false }
    ];
    localStorage.setItem('todos', JSON.stringify(todos));
    render(
      <I18nProvider i18n={i18n}>
        <App>
          <TodoList todos={todos}>
            <Todo />
          </TodoList>
        </App>
      </I18nProvider>
    );

    localStorage.setItem('todos', JSON.stringify(todos));

    //setItem('todos', JSON.stringify(todos));

    const addButton = screen.getByText('Add');
    expect(addButton).toBeInTheDocument();

    const textInput = screen.getByRole('textbox');
    expect(textInput).toBeInTheDocument();
    expect(textInput).toHaveValue('');

    const noText = screen.queryByText(/buy coconuts/i);
    expect(noText).toBeNull();

    // If I enter a alphanumeric text (/\S+/g)
    userEvent.type(textInput, 'buy coconuts');
    expect(textInput).toHaveValue('buy coconuts');
    userEvent.click(addButton);
    expect(textInput).toHaveValue('');

    // const todos = JSON.parse(localStorage.getItem('todos'))
    //expect(noText).not.toBeInTheDocument()
    const text = screen.queryByText(/buy coconuts/i);
    expect(text).toBeInTheDocument();

    const regex = /\S+/g;
    expect(regex.test('buy coconuts')).toBeTruthy();

    expect(todos[0].note).toBe('buy bananas');
    expect(todos[1].note).toBe('buy strawberries');
    //expect(todos[2].note).toBe('buy coconuts');

    // If I enter a non-alphanumeric text
    // userEvent.type(textInput, "")
    userEvent.type(textInput, '  ');
    expect(textInput).toHaveValue('  ');
    userEvent.click(addButton);
    expect(textInput).toHaveValue('');

    // const text2 = screen.queryByText("")
    // expect(text2).toBeInTheDocument()

    expect(regex.test('')).toBeFalsy();

    expect(todos[0].note).toBe('buy bananas');
    expect(todos[1].note).toBe('buy strawberries');
    //expect(todos[2].note).toBeInTheDocument();
  });

  // works
  test('saveData()', () => {
    let todos = [
      { id: 'a', note: 'buy bananas', completed: false },
      { id: 'b', note: 'buy strawberries', completed: false }
    ];
    localStorage.setItem('todos', JSON.stringify(todos));
    render(
      <I18nProvider i18n={i18n}>
        <App>
          <TodoList todos={todos}>{/*<Todo todo={todo} />{  */}</TodoList>
        </App>
      </I18nProvider>
    );
    // does not work:
    //jest.spyOn(localStorage, "setItem");
    //localStorage.setItem = jest.fn();

    // works:
    // this works to --> window.localStorage.__proto__.setItem = jest.fn();
    jest.spyOn(window.localStorage.__proto__, 'setItem');

    localStorage.clear();
    localStorage.setItem('todos', JSON.stringify(todos));

    let currentItems = JSON.parse(localStorage.getItem('todos'));
    //console.log(currentItems)

    expect(currentItems).toEqual(todos);
    localStorage.clear();
  });

  /* 
    Source: https://jestjs.io/docs/manual-mocks
  
  #  Mocking methods which are not implemented in JSDOM
  
  
  If some code uses a method which JSDOM (the DOM implementation used by Jest) hasn't implemented yet, testing it is not easily possible. This is e.g. the case with window.matchMedia(). Jest returns TypeError: window.matchMedia is not a function and doesn't properly execute the test.
  
  In this case, mocking matchMedia in the test file should solve the issue:
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
  
  This works if window.matchMedia() is used in a function (or method) which is invoked in the test. If window.matchMedia() is executed directly in the tested file, Jest reports the same error. In this case, the solution is to move the manual mock into a separate file and include this one in the test before the tested file:
  import './matchMedia.mock'; // Must be imported before the tested file
  import {myMethod} from './file-to-test';
  
  describe('myMethod()', () => {
    // Test the method here...
  });
  
     */
});
