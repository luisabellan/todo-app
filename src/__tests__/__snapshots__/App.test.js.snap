// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="app-wrapper"
    data-testid="app"
  >
    <div
      data-testid="header"
    >
      <h2>
        Todo App
      </h2>
    </div>
    <div
      class="main"
    >
      <div>
        <form
          data-testid="form"
        >
          <input
            class="input"
            data-testid="todo-input"
            name="todoItem1"
            placeholder="New Task"
            type="text"
            value=""
          />
          <div
            id="buttons"
          >
            <div
              id="button-left"
            >
              <button
                class="add-todo-btn"
                data-testid="add-todo-button"
              >
                <span>
                  Add
                </span>
              </button>
            </div>
            <div
              id="button-right"
            >
              <button
                class="clear-btn"
              >
                <span>
                  Completed
                </span>
              </button>
              <button
                class="del-btn"
              >
                <span>
                  Clear
                </span>
              </button>
            </div>
          </div>
        </form>
      </div>
      <div
        class="todolist-wrapper"
        data-testid="todos"
      >
        <div
          class="checkbox"
          data-testid="todo"
        >
          <input
            data-testid="checkbox buy bananas"
            name="item buy bananas"
            type="checkbox"
            value="buy bananas"
          />
          <span
            data-testid="todo-output"
          >
            buy bananas
          </span>
        </div>
        <div
          class="checkbox"
          data-testid="todo"
        >
          <input
            data-testid="checkbox buy strawberries"
            name="item buy strawberries"
            type="checkbox"
            value="buy strawberries"
          />
          <span
            data-testid="todo-output"
          >
            buy strawberries
          </span>
        </div>
      </div>
    </div>
    <div
      data-testid="footer"
    >
      <footer
        class="footer"
      >
        <span
          class="copyright"
        >
          Made with ðŸ’“ and â˜• by 
          <a
            href="https://luisabellan.com"
          >
            Luis Abellan
          </a>
        </span>
      </footer>
    </div>
  </div>
</DocumentFragment>
`;
